### ISSUE TITLE
PCL support going forward

### ISSUE THREAD

**Issue by chambbj** [PDAL/PDAL#1404]
PDAL has matured tremendously since the time we first started using PCL as a dependency. At the same time, PCL development, while not halting, has slowed significantly.

In speaking with PDAL users at conferences and other venues, we've heard both sides of the argument. Some folks think that PCL is too much of a hassle and would like to be able to move forward without it. At the same time, others are excited about the ability to plugin PCL algorithms with relative ease.

I think we should _reduce_ the PCL footprint, while not removing support altogether. Here are my initial lists of PCL stages/kernels to either **keep** or **deprecate** along with justification. And, if we can agree to deprecate some items, what should the timeline be? One or two releases from now?

### Keep

* `filters.pclblock`

    Takes a JSON pipeline of PCL operations and strings them together in a pipeline. Only requires that the requested PCL function be exposed in `PCLPipeline`.

    Allows us to do anything one of the other, dedicated PCL filters (e.g., `filters.voxelgrid`) could do.

    With PDAL now accepting JSON pipelines, there may be opportunities to revisit how these two JSON pipeline specifications work together.

* `readers.pcd`

    Some users continue to want to convert data to PCD. We could implement this natively (we had a partial implementation before), but it seems like a lot of work for very little payoff.

* `writers.pcd`

    Some users continue to want to convert data from PCD. We could implement this natively (we had a partial implementation before), but it seems like a lot of work for very little payoff.

### Deprecate

* `filters.dartsample`

    Has been replaced by `filters.sample`.

* `filters.ground`

    Has been replaced by `filters.pmf`, although I believe we have one customer still using `filters.ground` in production at the moment.

* `filters.height`

    Has been replaced by `filters.hag`.

* `filters.greedyprojection`, `filters.gridprojection`, and `filters.movingleastsquares`

    Surface reconstruction algorithms with no known users. Implemented out of curiosity, but with somewhat unsatisfactory results. Was it the data? Or the parameters? We're not sure.

* `filters.poisson`

    Don't confuse it with Poisson disk sampling (aka dart throwing, aka `filters.sample`). This is a Poisson surface reconstruction algorithm, that hasn't yielded very good results in any of the aerial lidar data I've looked at.

* `filters.radiusoutlier` and `filters.statisticaloutlier`

    Have been replaced by `filters.outlier`.

* `filters.voxelgrid`

    Users may be familiar with this approach. It is intuitive and can return a somewhat predictable (at least bounded) number of points. We've shown a number of times though that the Poisson disk sampling (see `filters.sample`) is more visually appealing anyway.

    You should still be able to run `VoxelGrid` through `filters.pclblock` without issue.

* `writers.pclvisualizer`

    Exists as a writer simply so that we can stick it on the end of a pipeline and visualize the results. But you probably have a tool to view the results anyway, and this doesn't write anything to disk (should you be happy with your result and want to save it).

* `kernels.ground`

    Easy enough to construct a pipeline, or use `pdal translate`:

```
pdal translate in.las out.las pmf
```

* `kernels.pcl`

    Easy enough to construct a pipeline, or use `pdal translate`:

```
pdal translate in.las out.las pclblock --filters.pclblock.filename=pipeline.json
```

* `kernels.smooth`

    A convenience kernel for running `filters.movingleastsquares`, which we are considering deprecating.

* `kernels.view`

   An easy way to use `writers.pclvisualizer` to view a point cloud. Again, many users have a tool to view their data anyway, and you could probably still use `pdal translate` to do this in an awkward way.

**Comment by hobu** [comment]
I think we should lean toward removal rather than deprecation to eliminate confusion and prevent people from starting to use things things more generally. The ones that are difficult or impossible to pull into PDAL we should leave, and if/when they are replaced, we can rename/drop the PCL-based filter at that time. Fewer drivers to manage in general, especially if they duplicate functionality in slightly different ways, seems like the best policy.

We will need to be vigilant with the release notes about these changes, and if we don't have a clear path forward for people because of the removal, we can revisit removal of the filter. 

Remove
---------

* `filters.height`
* `filters.ground`
* `filters.dartsample`
* `filters.radiusoutlier`
* `filters.statisticaloutlier`
* `writers.pclvisualizer`
* `kernels.ground`
* `kernels.view`

Deprecate
------------
* `filters.movingleastsquares`
* `filters.gridprojection`
* `filters.greedyprojection`
* `filters.voxelgrid`
* `filters.poisson`

`

**Comment by loganbyers** [comment]
I would like to see `readers.pcd` and `writers.pcd` kept around. It seems like the format is simple enough to implement native readers and writers. No opinion on the filters or kernels. I agree that `writers.pclvisualizer` could be removed.

**Comment by abellgithub** [comment]
Have we tested the native drivers to make sure that they produce the same results as the ones they replace?

**Comment by chambbj** [comment]
@abellgithub Some. Probably not enough. I'd expect more thorough comparison prior to pulling the plug.

**Comment by chambbj** [comment]
Closed via #1432
